Clarify ownership of runner components

larryliu0820 opened the pull request:
Summary:
The ownership of these components need some clarification.

* `Module` should be shared by `TextDecoderRunner` and potentially `TextPrefiller` (or `ImagePrefiller` in multimodal runner).
* `TextDecoderRunner` should be shared by the `TextPrefiller` and `TextTokenGenerator`.
* `Tokenizer` should be owned by the `Runner` as well as `TextTokenGenerator`.

Differential Revision: D73399600

swolchok commented on examples/models/llama/tokenizer/llama_tiktoken.cpp:
```diff
@@ -11,51 +11,85 @@
+// Compile-time special tokens selection using templates
+template <Version V>
+struct SpecialTokensSelector {
+  static std::unique_ptr<std::vector<std::string>> get();
```
I would call this `create`; `get` connotes that it's somehow returning a thing that already exists

swolchok commented on examples/models/llava/runner/llava_text_decoder_runner.h:
```diff
@@ -17,7 +17,8 @@ namespace example {
-  explicit LlavaTextDecoderRunner(executorch::extension::Module* module)
+  explicit LlavaTextDecoderRunner(
+      std::shared_ptr<executorch::extension::Module> module)
       : TextDecoderRunner(module, true) {}
```
need to `std::move(module)`

larryliu0820 replied:
Do we do std::move for shared_ptr as well? I always thought only unique_ptr uses it.

swolchok commented on extension/llm/runner/image_prefiller.h:
```diff
@@ -21,7 +21,8 @@ namespace llm {
-  explicit ImagePrefiller(::executorch::extension::Module* module)
+  explicit ImagePrefiller(
+      std::shared_ptr<::executorch::extension::Module> module)
       : module_(module) {}
```
ditto move

swolchok commented on extension/llm/runner/text_decoder_runner.cpp:
```diff
@@ -21,7 +21,9 @@ namespace llm {
+TextDecoderRunner::TextDecoderRunner(
+    std::shared_ptr<Module> module,
+    bool use_kv_cache)
     : module_(module), use_kv_cache_(use_kv_cache) {}
```
move

swolchok commented on extension/llm/runner/text_prefiller.cpp:
```diff
@@ -17,7 +17,7 @@ namespace extension {
 
 TextPrefiller::TextPrefiller(
-    TextDecoderRunner* text_decoder_runner,
+    std::shared_ptr<TextDecoderRunner> text_decoder_runner,
```
need to move this on line 24 below

swolchok commented on extension/llm/runner/text_token_generator.h:
```diff
@@ -21,17 +21,19 @@ namespace llm {
       std::unique_ptr<std::unordered_set<uint64_t>>&& eos_ids,
-      Stats* stats)
+      std::shared_ptr<Stats> stats)
       : tokenizer_(tokenizer),
         text_decoder_runner_(text_decoder_runner),
         eos_ids_(std::move(eos_ids)),
         use_kv_cache_(use_kv_cache),
```
move

swolchok reviewed (changes requested):
please also get the information about ownership from your summary into the code as comments so we can tell what things are supposed to be shared between

larryliu0820 commented:
> please also get the information about ownership from your summary into the code as comments so we can tell what things are supposed to be shared between

Instead of using shared_ptr I added comments to explain the ownership. Still use raw pointers for Tokenizer and TextDecoderRunner etc.

larryliu0820 commented on examples/models/llama/tokenizer/llama_tiktoken.cpp:
```diff
@@ -11,51 +11,71 @@
 
-std::unique_ptr<Tiktoken> get_tiktoken_for_llama(Version version) {
+namespace detail {
+// Helper function to create a Tiktoken with the given version
+template <Version V>
+std::unique_ptr<Tiktoken> create_tiktoken() {
+  std::unique_ptr<std::vector<std::string>> special_tokens =
+      example::SpecialTokensSelector<V>::create();
   return std::make_unique<Tiktoken>(
-      _get_special_tokens(version), kBOSTokenIndex, kEOSTokenIndex);
+      std::move(special_tokens), kBOSTokenIndex, kEOSTokenIndex);
+}
+} // namespace detail
```
The reason is I wanted to use this to support tokenizer for different models. I guess it doesn't belong to the current PR so I'll separate it out.

larryliu0820 committed:
Clarify ownership of runner components (#10338)

Summary:
Pull Request resolved: https://github.com/pytorch/executorch/pull/10338

The ownership of these components need some clarification.

* `Module` should be shared by `TextDecoderRunner` and potentially `TextPrefiller` (or `ImagePrefiller` in multimodal runner).
* `TextDecoderRunner` should be shared by the `TextPrefiller` and `TextTokenGenerator`.
* `Tokenizer` should be owned by the `Runner` as well as `TextTokenGenerator`.

Reviewed By: kirklandsign

Differential Revision: D73399600

larryliu0820 commented on examples/models/llama/runner/runner.h:
```diff
@@ -71,7 +71,7 @@ class ET_EXPERIMENTAL Runner : public executorch::extension::llm::IRunner {
 
   // stats
-  ::executorch::extension::llm::Stats stats_;
+  std::unique_ptr<::executorch::extension::llm::Stats> stats_;
```
Ah that's because TextTokenGenerator needs to share the same stats with Runner, TextTokenGenerator(stats) but also Runner constructor is going to take in a TextTokenGenerator. Runner(text_token_generator)

Therefore the only way I can think of is to:

```cpp
auto stats = std::make_unique<Stats>();
auto text_token_generator = std::make_unique<TextTokenGenerator>(stats.get());
auto runner = std::make_unique<Runner>(std::move(stats), std::move(text_token_generator));
```

larryliu0820 replied:
of course I can put this change into next PR.

swolchok replied:
if you want to avoid the heap allocation, you can have the Runner hand a pointer to its Stats object to the TextTokenGenerator, or you can have the runner create the TextTokenGenerator and leave the constructor as-is.

swolchok reviewed (approved):
LGTM
